lib/spack/spack/test/cmd_extensions.py

```diff
@@ -79,33 +131,141 @@ def hello(parser, args):
         hello_folks()
     elif args.subcommand == 'global':
         print(global_message)
-""")
-    implementation = extension_root.ensure('testcommand', 'implementation.py')
-    implementation.write("""
+""".format(ext_pname=extension.pname))
+
+            extension.main.ensure('__init__.py')
+            implementation \
+                = extension.main.ensure('implementation.py')
+            implementation.write("""
 def hello_world():
     print('Hello world!')
 
 def hello_folks():
     print('Hello folks!')
 """)
-    list_of_modules = list(sys.modules.keys())
-    with spack.config.override('config:extensions', [str(extension_root)]):
-        yield spack.main.SpackCommand('hello')
+            yield spack.main.SpackCommand('hello')
 
-    to_be_deleted = [x for x in sys.modules if x not in list_of_modules]
-    for module_name in to_be_deleted:
-        del sys.modules[module_name]
+    yield _hwwmir
 
 
 def test_simple_command_extension(hello_world_cmd):
+    """Basic test of a functioning command."""
     output = hello_world_cmd()
     assert 'Hello world!' in output
 
 
-def test_command_with_import(hello_world_with_module_in_root):
-    output = hello_world_with_module_in_root('world')
-    assert 'Hello world!' in output
-    output = hello_world_with_module_in_root('folks')
-    assert 'Hello folks!' in output
-    output = hello_world_with_module_in_root('global')
-    assert 'bar' in output
+def test_multi_extension_search(hello_world_extension, tmpdir):
+    """Ensure we can find an extension command even if it's not in the first
+    place we look.
+    """
+
+    extra_ext_name = 'testcommand2'
+    extra_ext = Extension(extra_ext_name,
+                          tmpdir.mkdir('spack-' + extra_ext_name))
+    with spack.config.override('config:extensions', [str(extra_ext.root)]):
+        assert ('Hello world') in spack.main.SpackCommand('hello-world')()
```

Comment:
Would it be possible to reduce the test to:
```suggestion
def test_multi_extension_search(hello_world_extension, extension_creator, tmpdir):
    """Ensure we can find an extension command even if it's not in the first
    place we look.
    """
    with extension_creator('testcommand2'):
        assert ('Hello world') in spack.main.SpackCommand('hello-world')()
```
now that we have a context manager?

Reply:
Good suggestion! In fact, I was even able to simplify a little further by removing the use of `tmpdir` as a fixture, per https://github.com/spack/spack/pull/13635/commits/2bf53b171b6a8effe2bfe33220a2acf28e7bc2c4. However, see https://github.com/spack/spack/pull/13635/commits/ab7045ffe5436dd39dcf1e0356d3ea4746966fe5 and #15312: steps need to be taken to ensure that if a test fails, a polluted configuration or module environment does not cause the failure of subsequent tests.

Reply:
The latest version looks good to me, thanks!